base/iterators.jl

```diff
@@ -1343,4 +1344,144 @@ only(x::NamedTuple) = throw(
+That is, the ratio of successive elements is constant, not the difference.
+
+!!! compat "Julia 1.7"
+    This function requires at least Julia 1.7.
+
+This is similar to `geomspace` in Python, and to `PowerRange` in Mathematica.
+
+# Examples
+```jldoctest
+julia> foreach(println, logrange(2 => 16, 4))
+2.0
+4.0
+8.0
+16.0
+
+julia> collect(logrange(1000 => 1, 7))
+7-element Vector{Float64}:
+ 1000.0
+  316.22776601683796
+  100.0
+   31.622776601683793
+   10.0
+    3.1622776601683795
+    1.0
+
+julia> ans ≈ 10 .^ (3:-0.5:0)
+true
+
+julia> collect(logrange(-1 => -2f0, 3))
+3-element Vector{Float32}:
+ -1.0
+ -1.4142135
+ -2.0
+
+julia> map(rad2deg∘angle, logrange(1 => -1+0im, 5))
+5-element Vector{Float64}:
+   0.0
+  45.0
+  90.0
+ 135.0
+ 180.0
+```
+"""
+function logrange(fromto::Pair{<:Number, <:Number}, len::Integer)
+    len >= 2 || throw(ArgumentError("logrange must have length of at least 2"))
+    start, stop = map(float, promote(fromto...))
+    ratio = ratio_nth_root(stop, start, Int(len)-1)
+    LogRange(start, ratio, Int(len))
+end
+
+ratio_nth_root(a::Number, b::Number, n::Int) = (a/b)^(1/n)
+ratio_nth_root(a::Float32, b::Float32, n::Int) = (Float64(a)/Float64(b))^(1/n)
+ratio_nth_root(a::ComplexF32, b::ComplexF32, n::Int) = (ComplexF64(a)/ComplexF64(b))^(1/n)
+function ratio_nth_root(a::T, b::T, m::Int) where {T<:Union{Float64, ComplexF64}}
+    over = Base.TwicePrecision(a) / Base.TwicePrecision(b)
+    r1 = Base.TwicePrecision((over.hi)^(1/m))
+    # Refine that using one step of Newton's method:
+    r1pow = prod(r1 for _ in 1:m-1)
+    r2 = (m-1)*r1/m + over / (m * r1pow)
+    # r2pow = prod(r2 for _ in 1:m-1)
+    # r3 = (m-1)*r2/m + over / (m * r2pow)
+end
+
+
+"""
+    LogRange{T,S}(start::T, ratio::S, len::Int)
+
+Iterator which multiplies by `ratio` at each step, giving in total `len` objects.
+Constructing this directly instead of using [`logrange`](@ref) avoids the cost of
+computing `ratio` to high precision.
+
+Note that its output is of type `T`, so iterating for intance `LogRange(2, pi, 10)` will
+lead to an `InexactError`, since `2*pi` cannot be converted to `Int`.
+
+See also: [`LinRange`](@ref Base.LinRange), [`range`](@ref Base.range), [`TwicePrecision`](@ref Base.TwicePrecision).
+
+# Examples
+
+```jldoctest
+julia> collect(LogRange(10,2,3))  # logrange(10 => 40, 3) would use Float64 not integers
+3-element Vector{Int64}:
+ 10
+ 20
+ 40
+
+julia> collect(LogRange(0.1, sqrt(10), 5))  # no corrections for rounding errors
+5-element Vector{Float64}:
+  0.1
+  0.316227766016838
+  1.0000000000000002
+  3.1622776601683804
+ 10.000000000000004
+
+julia> r = logrange(0.1 => 10, 5)  # type T != type S
+LogRange{Float64, Base.TwicePrecision{Float64}}(0.1, Base.TwicePrecision{Float64}(3.162277660168379, 2.0941562178568784e-16), 5)
+
+julia> collect(r)
+5-element Vector{Float64}:
+  0.1
+  0.31622776601683794
+  1.0
+  3.1622776601683795
+ 10.0
+```
+"""
+struct LogRange{T,S}
+    start::T
```

Comment:
Rather than start, perhaps you could parallel the `ref` and `offset` properties of `StepRangeLen`, which is by default the first index:
https://github.com/JuliaLang/julia/blob/05a80e1e82ac976bd3323f10e1139d86b3954cbe/base/range.jl#L331-L341

Perhaps you could parameterize `offset` as a `Val` which might signal whether you should use multiplication or exponentiation during iteration.

Reply:
I hope it won't be necessary to be as complicated as `StepRangeLen`! At the moment the plan is only to provide an iterator.